diff --git a/configure.ac b/configure.ac
index 36c24b4db..2241255b7 100644
--- a/configure.ac
+++ b/configure.ac
@@ -850,11 +850,10 @@ AC_ARG_WITH([util], AS_HELP_STRING([--without-util], [compile without libutil]),
 AS_IF([test "x$with_util" = xno], [
   AM_CONDITIONAL([HAVE_UTIL], [false])
   have_util=no
-], [
-  UL_CHECK_LIB([util], [openpty])
-])
+] )
+UL_CHECK_LIB([zoslib], [openpty])
 
-AS_IF([test "x$have_pty_h" = xyes && test "x$have_sys_signalfd_h" = xyes && test "x$have_util" = xyes], [
+AS_IF([test "x$have_pty_h" = xyes && test "x$have_sys_signalfd_h" = xyes && test "x$have_zoslib" = xyes], [
    AM_CONDITIONAL([HAVE_PTY], [true])
    AC_DEFINE([HAVE_PTY], [1], [have PTY support])
    have_pty=yes
@@ -2110,7 +2109,7 @@ AC_ARG_ENABLE([scriptutils],
 
 UL_ENABLE_ALIAS([script], [scriptutils])
 UL_BUILD_INIT([script])
-UL_REQUIRES_HAVE([script], [pty], [openpty function (libutil)])
+UL_REQUIRES_HAVE([script], [pty], [openpty function (zoslib)])
 AM_CONDITIONAL([BUILD_SCRIPT], [test "x$build_script" = xyes])
 
 UL_ENABLE_ALIAS([scriptreplay], [scriptutils])
@@ -2119,7 +2118,7 @@ AM_CONDITIONAL([BUILD_SCRIPTREPLAY], [test "x$build_scriptreplay" = xyes])
 
 UL_ENABLE_ALIAS([scriptlive], [scriptutils])
 UL_BUILD_INIT([scriptlive])
-UL_REQUIRES_HAVE([scriptlive], [pty], [openpty function (libutil)])
+UL_REQUIRES_HAVE([scriptlive], [pty], [openpty function (zoslib)])
 AM_CONDITIONAL([BUILD_SCRIPTLIVE], [test "x$build_scriptlive" = xyes])
 
 
