diff --git a/term-utils/scriptlive.c b/term-utils/scriptlive.c
index f8434ad57..1bf9e8d6a 100644
--- a/term-utils/scriptlive.c
+++ b/term-utils/scriptlive.c
@@ -130,7 +130,11 @@ static int process_next_step(struct scriptlive *ss)
 
 		/* no delay -- immediately write */
 		rc = replay_emit_step_data(ss->setup, ss->step, fd);
+#ifndef __MVS__
 		fdatasync(fd);
+#else
+		fsync(fd);
+#endif
 	} while (rc == 0);
 
 	return rc;
@@ -146,7 +150,11 @@ static int mainloop_cb(void *data)
 		int fd = ul_pty_get_childfd(ss->pty);;
 
 		rc = replay_emit_step_data(ss->setup, ss->step, fd);
-		fdatasync(fd);
+#ifndef __MVS__
+                fdatasync(fd);
+#else
+                fsync(fd);
+#endif
 		if (rc)
 			return rc;
 	}
