diff --git a/term-utils/scriptlive.c b/term-utils/scriptlive.c
index e4a3434..d400435 100644
--- a/term-utils/scriptlive.c
+++ b/term-utils/scriptlive.c
@@ -130,7 +130,11 @@ static int process_next_step(struct scriptlive *ss)
 
 		/* no delay -- immediately write */
 		rc = replay_emit_step_data(ss->setup, ss->step, fd);
+#ifndef __MVS__
 		fdatasync(fd);
+#else
+		fsync(fd);
+#endif
 	} while (rc == 0);
 
 	return rc;
@@ -146,7 +150,11 @@ static int mainloop_cb(void *data)
 		int fd = ul_pty_get_childfd(ss->pty);;
 
 		rc = replay_emit_step_data(ss->setup, ss->step, fd);
-		fdatasync(fd);
+#ifndef __MVS__
+                fdatasync(fd);
+#else
+                fsync(fd);
+#endif
 		if (rc)
 			return rc;
 	}
diff --git a/text-utils/hexdump.c b/text-utils/hexdump.c
index 3674455..15fc7f0 100644
--- a/text-utils/hexdump.c
+++ b/text-utils/hexdump.c
@@ -41,7 +41,9 @@
 #include <string.h>
 #include <stdlib.h>
 #include <errno.h>
+#ifndef __MVS__
 #include <err.h>
+#endif
 #include <limits.h>
 #include <getopt.h>
 #include "hexdump.h"
